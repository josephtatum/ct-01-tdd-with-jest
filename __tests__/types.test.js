const {
  isNumber,
  isString,
  isBoolean,
  isArray,
  isObject,
  isFunction,
  castToNumber,
  castToString,
  getCaster
} = require('../lib/types.js');

describe('validator module', () => {
  describe('basic validation', () => {
    it('properly tells if a value is a number', () => {
      expect(isNumber(3)).toBeTruthy();
      expect(isNumber('hi')).toBeFalsy();
      expect(isNumber([])).toBeFalsy();
      expect(isNumber({})).toBeFalsy();
      expect(isNumber(() => {})).toBeFalsy();
      expect(isNumber(true)).toBeFalsy();
    });

    it('properly tells if a value is a string', () => {
      expect(isString('hello')).toBeTruthy();
      expect(isString('4')).toBeTruthy();
      expect(isString('true')).toBeTruthy();
      expect(isString(4)).toBeFalsy();
      expect(isString(true)).toBeFalsy();
    });

    it('properly tells if a value returns a boolean', () => {
      expect(isBoolean(true)).toBeTruthy();
      expect(isBoolean(false)).toBeTruthy();
      expect(isBoolean('true')).toBeFalsy();
      expect(isBoolean('false')).toBeFalsy();
      expect(isBoolean(1)).toBeFalsy();
    });

    it('properly tells if a value is an array', () => {
      expect(isArray([1, 2, 3])).toBeTruthy();
      expect(isArray('array')).toBeFalsy();
      expect(isArray({})).toBeFalsy();
      expect(isArray(1)).toBeFalsy();
    });

    it('properly tells if a value is an object', () => {
      expect(isObject({ name: 'doug' })).toBeTruthy();
      expect(isObject([1, 2, 3])).toBeFalsy();
      expect(isObject('{}')).toBeFalsy();
      expect(isObject(1)).toBeFalsy();
      expect(isObject([{ name: 'doug' }])).toBeFalsy();
    });

    it('properly tells if a value is a function', () => {
      expect(isFunction(isNumber)).toBeTruthy();
      expect(isFunction(isNumber(3))).toBeFalsy();
      expect(isFunction('function')).toBeFalsy();
      expect(isFunction(1)).toBeFalsy();
      expect(isFunction(false)).toBeFalsy();
    });
  });

  describe('casters', () => {
    it('can cast values to a number', () => {
      expect(castToNumber(3)).toEqual(3);
      expect(castToNumber('3')).toEqual(3);
      expect(castToNumber(true)).toEqual(1);
      expect(castToNumber(false)).toEqual(0);
    });

    it('throws error if value is not castable to number', () => {
      expect(() => castToNumber('hi')).toThrowErrorMatchingSnapshot();
      expect(() => castToNumber({})).toThrowErrorMatchingSnapshot();
    });

    it('can cast values to a string', () => {
      expect(castToString(3)).toEqual('3');
      expect(castToString('3')).toEqual('3');
      expect(castToString(true)).toEqual('true');
      expect(castToString({})).toEqual('{}');
    });

    it('throws error if a function expression is passed to castToString()', () => {
      expect(() => castToString(isString)).toThrowErrorMatchingSnapshot();
      expect(() => castToString(undefined)).toThrowErrorMatchingSnapshot();
    });
  });

  it('can get the right caster', () => {
    expect(getCaster(Number)).toEqual(castToNumber);
    expect(getCaster(Promise)).toBeNull();
  });

});
