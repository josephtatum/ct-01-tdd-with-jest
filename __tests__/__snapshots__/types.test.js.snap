// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validator module casters throws error if a function expression is passed to castToString() 1`] = `
"Cannot cast >>val => {
  /* istanbul ignore next */
  cov_zrany366n.f[1]++;
  cov_zrany366n.s[3]++;
  return typeof val === 'string';
}<< to String"
`;

exports[`validator module casters throws error if a function expression is passed to castToString() 2`] = `"Cannot cast >>undefined<< to String"`;

exports[`validator module casters throws error if value is not castable to number 1`] = `"Cannot cast >>hi<< to Number"`;

exports[`validator module casters throws error if value is not castable to number 2`] = `"Cannot cast >>[object Object]<< to Number"`;
